FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+            byte queueType,
+            string[] definitions,
+            long? groupId,
+            bool forceOneActiveJobGroup,
+            bool isCompleted,
+            CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Start to enqueue {definitions.Length} jobs.");
+
+            // step 1: get incremental job ids, will try again if fails
+            var jobIds = await GetIncrementalJobIds(queueType, definitions.Length, cancellationToken);
+
+            // handle for each job
+            var jobInfos = new List<TJobInfo>();
+            for (var i = 0; i < definitions.Length; i++)
+            {
+                Trace.Assert(
+                    definitions[i].Length * sizeof(char) < MaximumSizeOfAnEntityInBytes,
+                    "The maximum size of a single table entity is 1MB, the size of definition is larger than 1MB.");
+
+                // step 2: create jobInfo entity and job lock entity
+                var newJobInfo = new TJobInfo
+                {
+                    Id = jobIds[i],
+                    QueueType = queueType,
+                    Status = JobStatus.Created,
+                    GroupId = groupId ?? 0,
+                    Definition = definitions[i],
+                    Result = string.Empty,
+                    CancelRequested = false,
+                    CreateDate = DateTime.UtcNow,
+                    HeartbeatDateTime = DateTime.UtcNow,
+                };
+
+                var jobInfoEntity = newJobInfo.ToTableEntity();
+
+                var jobLockEntity = new JobLockEntity
+                {
+                    PartitionKey = jobInfoEntity.PartitionKey,
+                    RowKey = AzureStorageKeyProvider.JobLockRowKey(newJobInfo.JobIdentifier()),
+                    JobInfoEntityRowKey = jobInfoEntity.RowKey,
+                };
+
+                // step 3: insert jobInfo entity and job lock entity in one transaction.
+                IEnumerable<TableTransactionAction> transactionAddActions = new List<TableTransactionAction>
+                {
+                    new (TableTransactionActionType.Add, jobInfoEntity),
+                    new (TableTransactionActionType.Add, jobLockEntity),
+                };
+
+                try
+                {
+                    _ = await _azureJobInfoTableClient.SubmitTransactionAsync(transactionAddActions, cancellationToken);
+                }
+                catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.AddEntityAlreadyExistsErrorCode))
+                {
+                    // Note: for cancelled/failed jobs, we don't allow resume it, will return the existing jobInfo.
+                    _logger.LogWarning(ex, "Failed to add entities, the entities already exist. Will fetch the existing jobs.");
+                }
+
+                // step 4: get the existing or newly added jobInfo entity and job lock entity
```

Comment:
Could you explain why we need to fetch the entity again here?

Reply:
We need to get `JobLockEntity` to get the ETag to update the job message info to `JobLockEntity`. 
We don't need to get `JobInfoEntity` here, I move it to the handling of  `AddEntityAlreadyExists` catch section.